catalog/cataloger_merge.go

```diff
@@ -128,6 +125,43 @@ func (c *cataloger) Merge(ctx context.Context, repository, leftBranch, rightBran
 	return mergeResult, err
 }
 
+func drainChannel(channel chan mergeBatchRecords) {
+	// If merge just exited on error, than the diffWorker could continue sending to channel until the buffer
+	// got full, and diffWorker would stay in memory, in the middle of a transaction - something that
+	// may create problems after a while.
+	// channel draining is needed so that diffWorker does not get stuck, and  can exit the transaction
+	for ok := true; ok; { // drain the channel, till it closes
+		_, ok = <-channel
+	}
+}
+
+func (c *cataloger) diffWorker(params doDiffParams, mergeBatchChan chan mergeBatchRecords, exitFlag *bool) {
+	_, _ = c.db.Transact(func(tx db.Tx) (interface{}, error) {
```

Comment:
General question - does the fact that we do diff in one isolation and apply the changes in the different one can it cause in-consistency issues? example: the diff find a conflict and we applied a change in a transaction we opened that there is a conflict already there.

Reply:
An ancient Chinese proverb say that if you think about it for a long enough  - you will find inconsistencies in transactions.

on the entries I see no problem. only committed entries are read from source(left Branch). committed entries are changed only by commit or merge (which is a special commit).
The read transaction takes a snapshot, so even if the source is committed while a merge is running - it will not be reflected.
there is a possibility that a source parent will be merged with grand parent while a merge is running (quite a strange situation). Then there may be a problem because the lineage of the parent will be updated without the diffWorker "seeing" those changes. Then when the merge finishes - It may read the new commited lineage. (depends on serialization mode. As long as we are in serializable mode - it has to happen in the millisecond between start of the merge transaction, and the start of the diffWorker)
we have 3 possible solutions:
1. read the parent lineage by diffWorker, and passing it to the merge.
2. Lock the possibility of cascading from-parent merges
3. Decide that since lineage will go the way of the Dodo - ignore it until we change architecture
I vote 3